File: archive/r/ruby/maximum-array-rotation.rb
# frozen_string_literal: true

USAGE = 'Usage: please provide a list of integers (e.g. "8, 3, 1, 2")'

def usage!
  warn USAGE
  exit 1
end

def parse_list(str)
  str.split(",").map { Integer(it.strip) }
rescue ArgumentError, NoMethodError
  usage!
end

def max_rotation_sum(arr)
  n = arr.size
  total_sum = arr.sum
  current = arr.each_with_index.sum { |v, i| i * v }

  max = current

  (1...n).each do |i|
    last = arr[n - i]
    current += total_sum - n * last
    max = current if current > max
  end

  max
end

raw = ARGV.first
usage! if raw.nil? || raw.strip.empty?

arr = parse_list(raw)

puts max_rotation_sum(arr)

File: archive/e/elixir/maximum-array-rotation.ex
defmodule MaximumArrayRotation do
  @doc """
  Returns the maximum weighted sum across all rotations of a list, where weighted sum
  is the sum of all the elements multiplied by their indices 
  """
  @spec maximum_array_rotation(arr :: list(integer())) :: integer()
  def maximum_array_rotation(arr) do
    sum = Enum.sum(arr)
    len = length(arr)

    initial =
      Enum.with_index(arr)
      |> Enum.map(fn {n, i} -> n * i end)
      |> Enum.sum()

    Enum.scan(arr, initial, &(&2 + &1 * len - sum)) |> Enum.max()
  end

  @spec parse_list(str :: String.t()) :: list(integer())
  def parse_list(""), do: []
  def parse_list(str), do: String.split(str, ", ") |> Enum.map(&String.to_integer/1)
end

with [arg] <- System.argv(),
     [_ | _] = arr <- MaximumArrayRotation.parse_list(arg) do
  MaximumArrayRotation.maximum_array_rotation(arr) |> IO.puts()
else
  _ -> IO.puts("Usage: please provide a list of integers (e.g. \"8, 3, 1, 2\")")
end
